File: src/main/java/ru/openbank/meetup/demo/java/slide6/AlsoUseCase.java
package ru.openbank.meetup.demo.java.slide6;

import lombok.extern.slf4j.Slf4j;
import org.springframework.util.StringUtils;

@Slf4j
public class AlsoUseCase {
    private boolean checkRegex(String field, ApplicationField fieldConfig) {
        String regex = fetchRegex(fieldConfig);
        if (regex == null) {
            log.trace("Regex is not specified for this field [{}]", fieldConfig.getName());
            return true;
        }
        return field.matches(regex);
    }

    private String fetchRegex(ApplicationField fieldConfig) {
        String regex = null;
        if (!StringUtils.isEmpty(fieldConfig.getRegEx())) {
            regex = fieldConfig.getRegEx();
        } else if (!StringUtils.isEmpty(fieldConfig.getType())) {
            regex = findRegExpByName(fieldConfig.getType());
        }
        return regex;
    }

    private String findRegExpByName(String type) {
        return "";
    }
}

File: src/main/kotlin/ru/openbank/meetup/demo/kotlin/slide6/AlsoUseCase.kt
package ru.openbank.meetup.demo.kotlin.slide6

import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

fun String?.regexMatches(fieldConfig: ApplicationField): Boolean {
    val regex = fetchRegex(fieldConfig)
    return when (regex) {
        null -> true.also { logger.trace("Regex is not specified for this field [${fieldConfig.name}]") }
        else -> this!!.matches(regex.toRegex())
    }
}

private fun fetchRegex(fieldConfig: ApplicationField): String? {
    return when {
        fieldConfig.regEx.isNotNullOrEmpty() -> fieldConfig.regEx
        fieldConfig.type.isNotNullOrEmpty() -> findRegExpByName(fieldConfig.type!!)
        else -> null
    }
}

private fun findRegExpByName(type: String): String? {
    return ""
}

private fun CharSequence?.isNotNullOrEmpty(): Boolean = !this.isNullOrEmpty()
